iso15118/shared/security.py

```diff
@@ -1112,6 +1115,128 @@ def decrypt_priv_key(
     raise DecryptionError()
 
 
+def derive_certificate_hash_data(certificate: bytes) -> Dict[str, str]:
+    """Extract certificate hash data to be used in an OCPP AuthorizeRequest.
+
+    Args:
+        certificate: The certificate in binary (DER) form.
+
+    Returns:
+        A dictionary with all information required for an OCSPRequestDataType
+        (2.36. OCSPRequestDataType, p. 382, OCPP 2.0.1 Part 2)
+
+    Raises:
+        CertAttributeError: if a certificate is provided with a hash algorithm
+            that OCPP doesn't accept.
+            Only SHA256, SHA384, and SHA512 are allowed.
+            (3.42 HashAlgorithmEnumType, p. 403, OCPP 2.0.1 Part 2)
+    """
+    certificate = load_der_x509_certificate(certificate)
+    public_key = certificate.public_key()
+    public_key_bytes = public_key.public_bytes(
+        encoding=Encoding.X962,
+        format=PublicFormat.UncompressedPoint,
+    )
+    # Per https://www.ibm.com/docs/en/i/7.2?topic=concepts-distinguished-name :
+    # Distinguished name (DN) is a term that describes the identifying information
+    # in a certificate and is part of the certificate itself.
+    # A certificate contains DN information for both the owner or requestor
+    # of the certificate (called the Subject DN) and the CA that issues the certificate
+    # (called the Issuer DN). Depending on the identification policy of the CA
+    # that issues a certificate, the DN can include a variety of information.
+    #
+    # Each CA has a policy to determine what identifying information the CA requires
+    # to issue a certificate. Some public Internet Certificate Authorities may require
+    # little information, such as a name and e-mail address.
+    # Other public CAs may require more information and require stricter proof of that
+    # identifying information before issuing a certificate.
+    #
+    # https://www.ibm.com/docs/en/ibm-mq/7.5?topic=certificates-distinguished-names
+    # provides more information about the attributes which may be included in a DN.
+    #
+    # In this case, we will get a name like:
+    # 'DC=MO,C=DE,O=Keysight Technologies,CN=PKI-1_CRT_MO_SUB2_VALID'
+    distinguished_name_bytes = certificate.issuer.public_bytes()
+    serial_number = certificate.serial_number
+
+    # Convert to the naming used in OCPP.
+    hash_algorithm_for_ocpp = certificate.signature_hash_algorithm.name.upper()
+    if hash_algorithm_for_ocpp not in {"SHA256", "SHA384", "SHA512"}:
+        raise CertAttributeError("Unknown hash algorithm")
+
+    public_key_hasher = Hash(certificate.signature_hash_algorithm)
+    public_key_hasher.update(public_key_bytes)
```

Comment:
the public key we should extract is the public key bytes from the Issuer certificate and not from the Contract Certificate

Reply:
Did this.  Additionally, may still move to the OCSPRequestBuilder here.